packages/unmock-core/src/service/state/transformers.ts

```diff
@@ -113,129 +195,53 @@ const spreadStateFromService = (
-              ? { ...scm, const: stateValue }
-              : typeof stateValue === "function"
-              ? { ...scm, "x-unmock-function": stateValue }
-              : null,
-        };
-        matches = { ...matches, ...spread };
-      } else if (hasNestedItems(scm) || isNonEmptyObject(scm)) {
-        debugLog(
-          `spreadStateFromService: Found ${key} in schema, traversing ${JSON.stringify(
-            scm,
-          )} and ${JSON.stringify(stateValue)}`,
-        );
-        // Option 3: Current scheme has matching key, state specifies an object - traverse schema and indirection
-        // `stateValue` at this point may also contain DSL elements, so we parse them before moving onwards
-        const { translated, cleaned } = DSL.translateDSLToOAS(stateValue, scm);
-        const spread = {
-          [key]: { ...spreadStateFromService(scm, cleaned), ...translated },
-        };
-        matches = {
-          ...matches,
-          ...oneLevelOfIndirectNestedness(scm, statePath, spread),
-        };
-      } else {
-        debugLog(
-          `spreadStateFromService: Found ${key} in schema, but more ` +
-            `traversal is needed and not possible -> missing value found`,
-        );
-        // Option 4: Current schema has matching key, but state specifies an object and schema has final value
-        matches[key] = null;
-      }
-    } else {
-      throw new Error(
-        `${statePath[key]} (object '${JSON.stringify(
-          statePath,
-        )}' with key '${key}') is not an object, string, number, boolean or function!`,
+  return Object.keys(statePath)
+    .map(key => {
+      debugLog(
+        `spreadStateFromService: traversing the given state, looking to match ${key}`,
       );
-    }
-  }
-  debugLog(
-    `spreadStateFromService: Results for this iteration: ${JSON.stringify(
-      matches,
-    )}`,
-  );
-  return matches;
+      const scm = serviceSchema[key];
+      const stateValue = statePath[key];
+      return scm === undefined
+        ? matchWhenMissingKey(serviceSchema, statePath, key)
+        : isConcreteValue(stateValue)
+        ? matchWithConcreteValue(scm, stateValue, key)
+        : matchWithNonConcreteValue(scm, statePath, key);
+    })
+    .reduce((obj, el) => ({ ...obj, ...el }), {});
 };
 
 // Items that hold nested contents in OAS
 const NESTED_SCHEMA_ITEMS = ["properties", "items", "additionalProperties"];
 
-const hasNestedItems = (obj: any) =>
-  NESTED_SCHEMA_ITEMS.some((key: string) => obj[key] !== undefined);
+const hasNoNestedItems = (obj: any) =>
+  NESTED_SCHEMA_ITEMS.every((key: string) => obj[key] === undefined);
 
 const isConcreteValue = (obj: any) =>
   ["string", "number", "boolean", "function"].includes(typeof obj);
 
-const isNonEmptyObject = (obj: any) =>
-  typeof obj === "object" && Object.keys(obj).length > 0;
+const isEmptyObject = (obj: any) =>
+  typeof obj !== "object" || Object.keys(obj).length === 0;
 
 const oneLevelOfIndirectNestedness = (
```

Comment:
TBH I have no idea what e.g. this method does :) Do we still need to have this many `any`s?

Reply:
The `oneLevelOfIndirectNestedness`?

Re `any`, most of them can probably be replaced with a `Schema` object, actually.

Reply:
Ah, the reason it was `any` is because `Schema` object does not define a `[key: string]: ...` signature. I'll just cast it to such with an intermediate type.